### DISCUSSION TITLE
How to define custom objects as query parameters in FastAPI (using PydanticV2 annotations)?

### DISCUSSION THREAD

**Discussion by DSamuylov** [fastapi/fastapi#10101] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import bson
from typing import Annotated

from pydantic import (
    field_validator,
    PlainSerializer,
    WithJsonSchema,
    BeforeValidator,
    TypeAdapter,
)
from fastapi import FastAPI


ObjectId = Annotated[
    bson.ObjectId,
    BeforeValidator(lambda x: bson.ObjectId(x) if isinstance(x, str) else x),
    PlainSerializer(lambda x: f"{x}", return_type=str),
    WithJsonSchema({"type": "string"}, mode="validation"),
    WithJsonSchema({"type": "string"}, mode="serialization"),
]

# Casting from str to ObjectId - WORKS
ta = TypeAdapter(
    ObjectId,
    config=dict(arbitrary_types_allowed=True),
)
ta.validate_python("5f7f9f0c2a0e0b0001d5b3d0")
ta.json_schema()

# Use ObjectId as a query parameter in FastAPI - FAILS:
app = FastAPI()

@app.get("/test")
def test(id: ObjectId) -> bool:
    return True
```

### Description

I am trying to figure out how to update my FastAPI app after from Pydantic v1 to Pydantic v2.

Before I had code that worked perfectly fine:

```
class ObjectId(bson.ObjectId):
    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, value):
        if isinstance(value, str):
            try:
                return bson.ObjectId(value)
            except bson.errors.InvalidId:
                raise ValueError("Invalid ObjectId")
        elif isinstance(value, bson.ObjectId):
            return value
        else:
            raise TypeError("ObjectId required")


app = FastAPI()

@app.get("/test")
def test(id: ObjectId) -> bool:
    return True
```

With the code that I provided above, I get the following error message:

> fastapi.exceptions.FastAPIError: Invalid args for response field! Hint: check that <class 'bson.objectid.ObjectId'> is a valid Pydantic field type. If you are using a return type annotation that is not a valid Pydantic field (e.g. Union[Response, dict, None]) you can disable generating the response model from the type annotation with the path operation decorator parameter response_model=None. Read more: https://fastapi.tiangolo.com/tutorial/response-model/

I am a bit confused, because it seems that the problem is not with a response model (which is bool in the example), but with a custom type in query parameter.

I also checked other suggested approaches, for example:

- [What is the new way to declare Mongo ObjectId with PyDantic v2.0^?](https://stackoverflow.com/questions/76686267/what-is-the-new-way-to-declare-mongo-objectid-with-pydantic-v2-0)
- [Using bson.ObjectId in Pydantic v2](https://stackoverflow.com/questions/76686888/using-bson-objectid-in-pydantic-v2/76722139#76722139)

There, it is suggested an alternative approach to introduce a custom annotation type, however, it still does not work for me.

I also did not find any relevant in the FastAPI documentation about query parameters and a list of Extra Data Types.

I would highly appreciate if you could help to find a possible solution.

### Operating System

macOS

### FastAPI Version

0.101.1

### Pydantic Version

2.1.1

### Python Version

Python 3.11.4

**Comment by romabozhanovgithub** [comment]
Hey, it seems like there is a problem. You can try this, but probably it is not, what you are looking for:
```python
import bson
from typing import Annotated

from pydantic import (
    PlainSerializer,
    WithJsonSchema,
    BeforeValidator,
)
from fastapi import FastAPI


ObjectId = Annotated[
    str,
    BeforeValidator(lambda x: bson.ObjectId(x) if isinstance(x, str) else x),
    PlainSerializer(lambda x: f"{x}", return_type=str),
    WithJsonSchema({"type": "string"}, mode="validation"),
    WithJsonSchema({"type": "string"}, mode="serialization"),
]

app = FastAPI()

@app.get("/test", response_model=None)
def test(id: ObjectId) -> bool:
    return True
```

**Comment by DSamuylov** [comment]
Hi @romabozhanovgithub, thank you a lot for your comment. Yes, unfortunately this solution does not work because `Annotated` should not change the type, and all code interpreters would assume that `ObjectId` here is an instance of `str`.

Hopefully @tiangolo would have a moment to confirm if it is a bug or misunderstanding about how to use a new version of Pydantic with FastAPI.

**Comment by UlfurOrn** [comment]
I was looking for an answer to the same question and found a similar discussion here: https://github.com/tiangolo/fastapi/discussions/10131

It in turn gave me a solution that worked for me. Hopefully it also works in your case:
```
ObjectId = Annotated[
    bson.ObjectId,
    BeforeValidator(lambda x: bson.ObjectId(x) if isinstance(x, str) else x),
    PlainSerializer(lambda x: f"{x}", return_type=str),
    WithJsonSchema({"type": "string"}, mode="validation"),
    WithJsonSchema({"type": "string"}, mode="serialization"),
    Query(),
]
```
or
```
@app.get("/test", response_model=None)
def test(id: Annotated[ObjectId, Query()]) -> bool:
    return True
```

I think the docs may benefit from adding an example for this usecase 🤔

**Reply by baktun95827** [reply]
This exactly solved my issue!

**Comment by DSamuylov** [comment]
Hi @UlfurOrn, thank you a lot for your suggestion, but for me the code still fails with:

> fastapi.exceptions.FastAPIError: Invalid args for response field! Hint: check that <class 'bson.objectid.ObjectId'> is a valid Pydantic field type. If you are using a return type annotation that is not a valid Pydantic field (e.g. Union[Response, dict, None]) you can disable generating the response model from the type annotation with the path operation decorator parameter response_model=None. Read more: https://fastapi.tiangolo.com/tutorial/response-model/

Does it run for you?

**Reply by DSamuylov** [reply]
Actually I am not sure that the code from https://github.com/tiangolo/fastapi/discussions/10131 could be adjusted to my case, because `MyNumber` is a child of `int` which is a [supported data type](https://fastapi.tiangolo.com/tutorial/extra-data-types/).

**Reply by JavierSanchezCastro** [reply]
Check this: https://devpress.csdn.net/mongodb/6303c11dc67703293080a14d.html

**Reply by DSamuylov** [reply]
Hi @JavierSanchezCastro, thank you for the link, but that discussion is about how to do it with Pydantic V1 and this is what I had before trying to update to Pydantic V2. Unfortunately, it does not answer the question.

**Comment by DSamuylov** [comment]
Anybody has an idea on how to fix it? It is still a problem...

**Reply by omi-donlimit** [reply]
Same here. Did you find any workaround in the meantime?

**Comment by Kludex** [comment, marked as answer]
- Created https://github.com/tiangolo/fastapi/issues/10259.
